### PR TITLE
Settings UI: hide settings for unavailable modules

### LINKED ISSUE #5212: Admin: Not able to hide security modules in new UI with jetpack_get_available_modules
A user in the forums reported that he's unable to remove VaultPress from the new UI using the `jetpack_get_available_modules` filter:

> That functionality is no longer working for for a couple items under the Security Tab (security scanning and site backups specifically) in the new React interface. Previously I could block “vaultpress” in the modules list using that code, but it’s not working anymore. Is there a different module name in the latest version?

Original report:
https://wordpress.org/support/topic/hide-modules-in-react-interface/

### PR DESCRIPTION
**Pull request by eliorivero**
Hide settings for all modules that are disabled through the `jetpack_get_available_modules` filter.

Fixes #5212

#### Changes proposed in this Pull Request:

- update reducer to initialize once, resolve ambiguity for searchTerm, optimize string matching, fail asap and return
- hide settings for disabled modules. Consolidate calls to avoid running isModuleFound twice since it's expensive.

#### Testing instructions:

* add code to disable modules and check that it's reflected in settings
```
add_filter( 'jetpack_get_available_modules', function( $modules ) {
	unset( $modules['masterbar'] );
	unset( $modules['after-the-deadline'] );
	unset( $modules['carousel'] );
	return $modules;
} );
```

### CODE REVIEW

File: _inc/client/writing/theme-enhancements.jsx [hunk 1]
```diff
@@ -98,130 +98,140 @@ const ThemeEnhancements = moduleSettingsForm(
...
-								<SettingsGroup hasChild module={ { module: item.module } } key={ `theme_enhancement_${ item.module }` } support={ item.learn_more_button }>
-									<FormLegend className="jp-form-label-wide">
+								return (
+									<SettingsGroup hasChild module={ { module: item.module } } key={ `theme_enhancement_${ item.module }` } support={ item.learn_more_button }>
+										<FormLegend className="jp-form-label-wide">
+											{
+												item.name
+											}
+										</FormLegend>
+>>>>>>> Settings UI - Writing: hide settings for disabled modules. Consolidate calls to avoid running isModuleFound twice since it's expensive.
```
Comment (dereksmart) [inline _inc/client/writing/theme-enhancements.jsx]: Failed rebase?

File: _inc/client/components/navigation-settings/index.jsx [hunk 2]
```diff
@@ -78,55 +83,102 @@ export const NavigationSettings = React.createClass( {
...
+	 * @param  {array}   modules Modules that are probably included in the module list.
+	 *
+	 * @return {boolean}         True if at least one of the modules is included in the list.
 	 */
-	buildUrl: function( href, keyword ) {
-		const splitUrl = href.split( '#' ),
-			splitHash = splitUrl[ 1 ].split( '?' );
-
-		this.props.searchForTerm( keyword );
-		return '#' + splitHash[ 0 ] + ( keyword ? '?term=' + keyword : '' );
```
Comment (dereksmart) [inline _inc/client/components/navigation-settings/index.jsx]: I think removing this broke search.  It refreshes the page and adds params weird places, for example searching for google results in `&s=google#/settings`

**Comment by dereksmart** [comment]
Works well other than my comments above

**Comment by eliorivero** [comment]
All updated and rebased. Thanks for reviewing.

**Comment by dereksmart** [comment]
No matter what the terms are, when searching, VaultPress and Akismet both appear in the results 

screen shot 2017-03-27 at 10 51 14 pm

File: _inc/client/security/index.jsx [hunk 3]
```diff
@@ -31,54 +30,58 @@ export const Security = React.createClass( {
...
-		);
 		return (
 			<div>
 				<QuerySite />
-				<QuerySitePlugins />
-				{ ( found.protect || found.sso ) && backupSettings }
-				{ ( found.protect || found.sso ) && akismetSettings }
-				{ found.protect && protectSettings }
-				{ found.sso && ssoSettings }
+				{
```
Comment (zinigor) [inline _inc/client/security/index.jsx]: This works fine in cases of filtered out modules, but has one caveat when you search. You can see in the previous versions that Backups and Akismet were only displayed when either protect or sso were "found". That prevented those cards from showing up in every search, as they do now. 

If we revert to previous logic, that problem would be solved, but we will have another problem - Backup and Akismet will disappear when Protect and SSO are both filtered out. But having Protect and SSO filtered out is a much less frequently happening edge case.

In the future we'll make sure we show and hide things based on features, not modules, and this problem will be solved entirely.

Comment (zinigor) [inline _inc/client/security/index.jsx]: Oh, I can see Derek already mentioned that problem.

**Comment by eliorivero** [comment]
@dereksmart @zinigor I pushed a fix for the issue in https://github.com/Automattic/jetpack/pull/6773/commits/ce640e6900b9a460a80853b5206d5ebe7c17559f
Akismet and VaultPress will now only show up in relevant searches.

**Pull request merged**